lib/fs.ts

```diff
@@ -34,12 +34,46 @@ interface PendingUpdate {
 	pendingDataFetcher: (() => string | Buffer) | null;
 	pendingOptions: object | null;
 	throttleTime: number; // throttling until time (0 for no throttle)
-	throttleTimer: {promise: Promise<void>, clear: () => void} | null;
+	throttleTimer: CancellableTimerPromise<void> | null;
 	finish: Promise<void>;
 }
 
 const pendingUpdates = new Map<string, PendingUpdate>();
 
+/**
+ * Givn an async function and ms, this will create a promise that resolves
+ * when the timer and async function ends
+ *
+ * It also has a `cancel` function that will clear the timer
+ * and resolve the promise
+ */
+export class CancellableTimerPromise<T, U extends T | PromiseLike<T> | undefined = undefined> {
```

Comment:
I wonder if this could extend `Promise`.

Reply:
It should be able to, but I don't see _why_ it should extend it.